### PR TITLE
New Resource: `azurerm_custom_ip_prefix`

### PR DESCRIPTION
**Pull request by sinbai**
Support for new resource azurerm_custom_ip_prefix. A few things need to be clarified.

1. Test cases are executed sequentially because the test IP range can only create one resource at a time. The total time for all the IPv4 tests is approximately 18 hours.

2. If the test case fails in some reason (e.g. `ResourceGroupBeingDeleted`), and the `commissionstate` of custom ip prefix in not in `Deprovisioned` or `ValidationFailed`, the custom ip prefix could not be deleted.  It should be updated to `Deprovisioned`(either in Azure portal by clicking deprovision or use powershell). Otherwise, the CIDR used in the custom ip prefix will not be able to be used to recreate custom ip prefix resource. The following is the relevant information about  commissionstate confirmed with service team:
```
-  `Provisioning`:    can do nothing
-  `ValidatedFailed`: can retry provision(backend start from the beginning, do WAN validation again) or delete
-  `Provisioned`:     can do commission,or deprovision
-  `ProvisionFailed`: can retry provision(backend skip WAN validation)
-  `Commissioning`:  can do nothing
-  `Commissioned`:   can do decommission
-  `CommissionFailed`: can do commission
-  `Decommissioning`:  can donothing （back to provisioned）
-  `DecommissionFailed`:can do decommission
-  `Deprovisioning`:  can do nothing
-  `Deprovisioned`:  can do delete or provision
-  `DeprovisionFailed`: can do deprovisioning
```

All tests passed in TeamCity.

### CODE REVIEW

File: internal/services/network/custom_ip_prefix_resource.go
Comment (tombuildsstuff) [inline internal/services/network/custom_ip_prefix_resource.go:106]: why is this exposed? isn't this the internal state?

File: internal/services/network/custom_ip_prefix_resource.go
Comment (tombuildsstuff) [inline internal/services/network/custom_ip_prefix_resource.go:153]: ... shouldn't we decommission, update and then recommission here?

File: internal/services/network/custom_ip_prefix_resource.go
Comment (tombuildsstuff) [inline internal/services/network/custom_ip_prefix_resource.go:306]: this timeout is being set within the `prepareDelete` method too - so by this point in time the timeout won't be correct, we'll need to pass `ctx` into `prepareDelete` (or remove that method)

File: website/docs/r/custom_ip_prefix.html.markdown
Comment (katbyte) [inline website/docs/r/custom_ip_prefix.html.markdown:54]: could we be more explicit in the name as what this should be
```suggestion
* `wan_authorization_message` - (Optional) The authorization message for WAN validation. Changing this forces a new resource to be created.
```

and given it is composed of 3 individual parts could we break it out into a block?

File: website/docs/r/custom_ip_prefix.html.markdown
Comment (katbyte) [inline website/docs/r/custom_ip_prefix.html.markdown:58]: could this be
```suggestion
* `wan_validation_signed_message` - (Optional) The signed message for WAN validation. Changing this forces a new resource to be created.
```

File: internal/services/network/custom_ip_prefix_resource_test.go
Comment (tombuildsstuff) [inline internal/services/network/custom_ip_prefix_resource_test.go:18]: since this is creating an IP Prefix, can we create a resource *from* this Resource? that would allow us to test this works end-to-end

File: internal/services/network/custom_ip_prefix_resource.go
Comment (katbyte) [inline internal/services/network/custom_ip_prefix_resource.go:46]: could we split this into separate create and update
```suggestion
		Create: resourceCustomIpPrefixCreate,
		Read:   resourceCustomIpPrefixRead,
		Update: resourceCustomIpPrefixUpdate,
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @sinbai 

Thanks for this PR - I've taken a look through and left some comments inline, but if we can fix those up then we can take another look - but we also need to ensure that we have some end-to-end tests of this resource being used.

Thanks!

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by katbyte** [comment]
> Due to an IPv6 issue in the API, currently only IPv4 is supported and confirmed with the service team that it cannot be fixed before GA (2022/03/31). The next service API release will support IPv6.
this GA date4 has passed so presumably we can implement ipv6?

> The test cases need to be run under the "eastus2euap" region, as the test IP range (provided by the service team for terraform testing) is now only available in that region.
Can we run the tests in a real region now?

> Test cases are executed sequentially because the test IP range can only create one resource at a time. The total time for the IPv4 test is approximately 7 hours.
can we use different IP ranges to speed this up now?

File: internal/services/network/custom_ip_prefix_resource.go
Comment (sinbai) [reply internal/services/network/custom_ip_prefix_resource.go:153]: Thanks for your feedback. For the [commissionedState](https://github.com/Azure/azure-rest-api-specs/blob/1b301846a0e31a668c59b264c900be2128762fac/specification/network/resource-manager/Microsoft.Network/stable/2021-08-01/customIpPrefix.json#L386) property,  there are "Provisioning",   "Provisioned", "Commissioning","Commissioned", "Decommissioning",Deprovisioning" values are listed in the v2021-0-01 API. Per the following information confirmed with service team, I added the check code above. I am not sure what the comment means, so could you please explain about it?

Provisioning: can do nothing  
Commissioning: can do nothing 
Decommissioning: can do nothing
Deprovisioning: can do nothing  
Provisioned: can do commission, or deprovision  
Commissioned: can do decommission  
Deprovisioned: can do delete or provision(backend start from the beginning, do WAN validation again)  

Note: If the action of `Decommission` is done, the commissionState back to Provisioned.

Thanks,
Elena

File: internal/services/network/custom_ip_prefix_resource_test.go
Comment (sinbai) [reply internal/services/network/custom_ip_prefix_resource_test.go:18]: @tombuildsstuff  Thanks for you feedback. For the comment above, do you mean to add a test case to create a `azurerm_public_ip_prefix` from this resource `azurerm_custom_ip_prefix`? If so, that is another feature requested from service team. That is "[Create a public IP prefix from a custom IP prefix](https://docs.microsoft.com/en-us/azure/virtual-network/ip-services/manage-custom-ip-address-prefix#create-a-public-ip-prefix-from-a-custom-ip-prefix)". Since this feature requires Terraform to support the [ip_prefix](https://github.com/hashicorp/terraform-provider-azurerm/blob/10a2ed50f79f1789320e76b775f38abd23d2c86b/internal/services/network/public_ip_prefix_resource.go#L85) of `azurerm_public_ip_prefix` setting, I plan to implement support it after this PR is merged. The end-to-end test  about creating a resource from this Resource would be added to `azurerm_public_ip_prefix` test at that time. WDYT?

File: internal/services/network/custom_ip_prefix_resource.go
Comment (sinbai) [reply internal/services/network/custom_ip_prefix_resource.go]: The commissionedState is not internal state since it could be updated after the custom ip prefix resource is created successfully.

File: internal/services/network/custom_ip_prefix_resource.go
Comment (sinbai) [reply internal/services/network/custom_ip_prefix_resource.go]: Code has been updated. Could you please see if this is what you mean?

File: internal/services/network/custom_ip_prefix_resource.go
Comment (sinbai) [reply internal/services/network/custom_ip_prefix_resource.go]: Fixed.

File: website/docs/r/custom_ip_prefix.html.markdown
Comment (sinbai) [reply website/docs/r/custom_ip_prefix.html.markdown]: Fixed.

File: website/docs/r/custom_ip_prefix.html.markdown
Comment (sinbai) [reply website/docs/r/custom_ip_prefix.html.markdown]: The `authorization_message` consists of subscription Id, cidr and roa expiration date. Now that the `cidr` property is  exposed, and the subs id must also be the subs id that created the custom ip prefix resource, maybe just exposing the `roa_expiration_date` property is enough, WDYT? Any feedback feel free let me know, thanks:).

File: internal/services/network/custom_ip_prefix_resource.go
Comment (sinbai) [reply internal/services/network/custom_ip_prefix_resource.go]: Fixed.

**Comment by sinbai** [comment]
@katbyte answers inline.
> 



> > Due to an IPv6 issue in the API, currently only IPv4 is supported and confirmed with the service team that it cannot be fixed before GA (2022/03/31). The next service API release will support IPv6.
> 
> this GA date4 has passed so presumably we can implement ipv6?

Per [doc](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/custom-ip-address-prefix#limitations) and reconfirmed by the service team, custom IP prefix does not currently support IPv6 for some reason.

> 
> > The test cases need to be run under the "eastus2euap" region, as the test IP range (provided by the service team for terraform testing) is now only available in that region.
> 
> Can we run the tests in a real region now?

yes, the tests could now be run in a real region, the code has been updated to remove the `eastus2euap` region restriction.

> 
> > Test cases are executed sequentially because the test IP range can only create one resource at a time. The total time for the IPv4 test is approximately 7 hours.
> 
> can we use different IP ranges to speed this up now?

After confirming with the service team, I am sorry to say it is not possible as they don't have more IP ranges could be used.

**Comment by katbyte** [comment]
@sinbai  - 7 hours for all tests or each test?

and we have test failures:

```

------- Stdout: -------
=== RUN   TestAccCustomIpPrefix_ipv4Update_from_deprovisioned_todo_decommission
    testcase.go:117: Step 1/6 error: Error running apply: exit status 1
        
        Error: waiting for Custom Ip Prefix: (Custom Ip Prefixe Name "acctest-CustomIpPrefix-220928174322198272" / Resource Group "acctestRG-auto-220928174322198272") to become ready: unexpected state 'ValidationFailed', wanted target 'Provisioned'. last error: %!s(<nil>)
        
          with azurerm_custom_ip_prefix.test,
          on terraform_plugin_test.tf line 11, in resource "azurerm_custom_ip_prefix" "test":
          11: resource "azurerm_custom_ip_prefix" "test" {
        
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: deleting Resource Group "acctestRG-auto-220928174322198272": the Resource Group still contains Resources.
        
        Terraform is configured to check for Resources within the Resource Group when deleting the Resource Group - and
        raise an error if nested Resources still exist to avoid unintentionally deleting these Resources.
        
        Terraform has detected that the following Resources still exist within the Resource Group:
        
        * `/subscriptions/*******/resourceGroups/acctestRG-auto-220928174322198272/providers/Microsoft.Network/customIpPrefixes/acctest-CustomIpPrefix-220928174322198272`
        
        This feature is intended to avoid the unintentional destruction of nested Resources provisioned through some
        other means (for example, an ARM Template Deployment) - as such you must either remove these Resources, or
        disable this behaviour using the feature flag `prevent_deletion_if_contains_resources` within the `features`
        block when configuring the Provider, for example:
        
        provider "azurerm" {
          features {
            resource_group {
              prevent_deletion_if_contains_resources = false
            }
          }
        }
        
        When that feature flag is set, Terraform will skip checking for any Resources within the Resource Group and
        delete this using the Azure API directly (which will clear up any nested resources).
        
        More information on the `features` block can be found in the documentation:
        https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#features
        
        
--- FAIL: TestAccCustomIpPrefix_ipv4Update_from_deprovisioned_todo_decommission (780.79s)
FAIL
```

**Comment by sinbai** [comment]
> TestAccCustomIpPrefix_ipv4Update_from_deprovisioned_todo_decommission

@katbyte It takes about 18 hours for all the test cases to run through. In my tests, the time spent in each case is as follows.

image
image
image
image

I updated the description of this PR as follows, the reason for failure of TestAccCustomIpPrefix_ipv4Update_from_deprovisioned_todo_decommission should be that Cidr is used by an existing custom ip prefix. This situation is what I mentioned in "If the test case fails in some reason..." below.
```
Support for new resource azurerm_custom_ip_prefix. A few things need to be clarified.

Test cases are executed sequentially because the test IP range can only create one resource at a time. The total time for all the IPv4 tests is approximately 18 hours.

If the test case fails in some reason (e.g. ResourceGroupBeingDeleted), and the commissionstate of custom ip prefix in not in Deprovisioned or ValidationFailed, the custom ip prefix could not be deleted. It should be updated to Deprovisioned(either in Azure portal by clicking deprovision or use powershell). Otherwise, the CIDR used in the custom ip prefix will not be able to be used to recreate custom ip prefix resource. The following is the relevant information about commissionstate confirmed with service team:

-  `Provisioning`:    can do nothing
-  `ValidatedFailed`: can retry provision(backend start from the beginning, do WAN validation again) or delete
-  `Provisioned`:     can do commission,or deprovision
-  `ProvisionFailed`: can retry provision(backend skip WAN validation)
-  `Commissioning`:  can do nothing
-  `Commissioned`:   can do decommission
-  `CommissionFailed`: can do commission
-  `Decommissioning`:  can donothing （back to provisioned）
-  `DecommissionFailed`:can do decommission
-  `Deprovisioning`:  can do nothing
-  `Deprovisioned`:  can do delete or provision
-  `DeprovisionFailed`: can do deprovisioning
```

File: website/docs/r/custom_ip_prefix.html.markdown
Comment (GurliGebis) [inline website/docs/r/custom_ip_prefix.html.markdown:29]: "singed", shouldn't that be "signed" ?

Comment (sinbai) [reply website/docs/r/custom_ip_prefix.html.markdown]: @GurliGebis thanks for your feedback. Fixed.

Comment (GurliGebis) [reply website/docs/r/custom_ip_prefix.html.markdown]: You're welcome 🙂

**Comment by brianlehr** [comment]
Hi team, how close are we to being complete with this?

**Comment by tombuildsstuff** [comment]
Pushing a rebase of this

**Comment by brianlehr** [comment]
> Pushing a rebase of this

@tombuildsstuff  Are we almost complete with this?  It has been open for 9 months at this point.

**Comment by tombuildsstuff** [comment]
hey @sinbai

Thanks for this PR - apologies for the delayed review of this one, I thought I'd left an update before the Christmas break.

As mentioned previously I've been working to try and review this PR, unfortunately I think there's an issue with the Service which I've been trying to diagnose where a Custom IP Prefix doesn't appear to be fully released until sometime after it's deleted.

Since the behaviour of the API has also changed considerably since this PR was first opened, supporting both specifying a CIDR block and (appearing to) provision one for you - I believe there maybe some other behavioural changes needed to this resource too, namely removing the `action` field in favour of taking a different approach. Whilst I've got local changes to do that - unfortunately I'm blocked on testing this at this point in time, since whilst we're running the tests for Custom IP Prefix, it appears that the CIDR block used by the Custom IP Prefix isn't fully reusable until sometime after the Custom IP Prefix is deleted.

Whilst I've been trying to diagnose that (by deleting any spurious Custom IP Prefixes, and waiting a couple of days  [since an hour didn't make much difference] before re-running the test to give us enough time to be sure there's an issue or not) - unfortunately an automated process keeps kicking these tests off, which is causing the need to start over - as has happened again this morning.

To workaround this, whilst I've got some experimental changes to this resource locally (as outlined above) - I'm going to push a commit to this branch to force-fail the tests for the moment, which should stop these creating Custom IP Prefixes temporarily until we can diagnose this issue. I've also reached out to ask that the automated process is stopped, however pushing this failure mode should be sufficient to catch it even if it is.

I've also removed the remaining Custom IP Prefixes within our account, so I'm hopeful that later in the week we should be finally able to diagnose this issue and determine how to proceed here.

As such, whilst I'd like to apologies for the delay reviewing this PR (and that I've forgotten to post this update before the Christmas break) - I'm hopeful that we should be able to diagnose this issue later in the week and then determine what's needed to proceed here.

Thanks,
Tom

**Comment by tombuildsstuff** [comment]
(Adding `waiting-response` to avoid the automated tooling too)

**Comment by brianlehr** [comment]
> hey @sinbai
> 
> Thanks for this PR - apologies for the delayed review of this one, I thought I'd left an update before the Christmas break.
> 
> As mentioned previously I've been working to try and review this PR, unfortunately I think there's an issue with the Service which I've been trying to diagnose where a Custom IP Prefix doesn't appear to be fully released until sometime after it's deleted.
> 
> Since the behaviour of the API has also changed considerably since this PR was first opened, supporting both specifying a CIDR block and (appearing to) provision one for you - I believe there maybe some other behavioural changes needed to this resource too, namely removing the `action` field in favour of taking a different approach. Whilst I've got local changes to do that - unfortunately I'm blocked on testing this at this point in time, since whilst we're running the tests for Custom IP Prefix, it appears that the CIDR block used by the Custom IP Prefix isn't fully reusable until sometime after the Custom IP Prefix is deleted.
> 
> Whilst I've been trying to diagnose that (by deleting any spurious Custom IP Prefixes, and waiting a couple of days [since an hour didn't make much difference] before re-running the test to give us enough time to be sure there's an issue or not) - unfortunately an automated process keeps kicking these tests off, which is causing the need to start over - as has happened again this morning.
> 
> To workaround this, whilst I've got some experimental changes to this resource locally (as outlined above) - I'm going to push a commit to this branch to force-fail the tests for the moment, which should stop these creating Custom IP Prefixes temporarily until we can diagnose this issue. I've also reached out to ask that the automated process is stopped, however pushing this failure mode should be sufficient to catch it even if it is.
> 
> I've also removed the remaining Custom IP Prefixes within our account, so I'm hopeful that later in the week we should be finally able to diagnose this issue and determine how to proceed here.
> 
> As such, whilst I'd like to apologies for the delay reviewing this PR (and that I've forgotten to post this update before the Christmas break) - I'm hopeful that we should be able to diagnose this issue later in the week and then determine what's needed to proceed here.
> 
> Thanks, Tom

FYI @gitlwh

@tombuildsstuff  - Curious to know what Custom IP Prefixes you are testing?  I ask as the PM for the product - we gave @sinbai  access to a single working Custom IP Prefix but above it references multiple?

**Comment by tombuildsstuff** [comment]
@brianlehr 

> @tombuildsstuff - Curious to know what Custom IP Prefixes you are testing? I ask as the PM for the product - we gave @sinbai access to a single working Custom IP Prefix but above it references multiple?

We're using `194.41.20.0/24` (as per the acctests in this branch), with each test running sequentially - however as mentioned above we're unable to successfully provision this at the moment as outlined above (we get a Validation Failed: CIDR in use - with no Custom IP Prefixes present in our Subscription), which is why I'm wondering if there's something kicking around here?

**Comment by sinbai** [comment]
@tombuildsstuff I have manually tested the IP range 194.41.20.0/24 and it could be successfully provisioned, deprovisioned and deleted completely. Could you take another look?

**Comment by manicminer** [comment]
Closing in favour of #21322

**Pull request closed without merging**
